temfpy/econometrics.py

```diff
@@ -0,0 +1,286 @@
+                for name in var_names
+            ]
+
+        start_params = bethas
+
+    else:
+        covariance = np.eye(n_choices - 1)
+        cov = []
+        for i in range(n_choices - 1):
+            for j in range(n_choices - 1):
+                if j <= i:
+                    cov.append(covariance[i, j])
+
+        cov = np.asarray(cov)
+
+        index_tuples = []
+        var_names = list(x.columns)
+        for choice in range(n_choices - 1):
+            index_tuples += [
+                ("choice_{}".format(choice), "betha_{}".format(name))
+                for name in var_names
+            ]
+
+        j = (n_choices) * (n_choices - 1) / 2
+        index_tuples += [("covariance", i) for i in range(int(j))]
+
+        start_params = np.concatenate((bethas, cov))
+
+    params_sr = pd.Series(
+        data=start_params, index=pd.MultiIndex.from_tuples(index_tuples), name="value"
+    )
+
+    y = y - y.min()
+
+    return (
+        y.to_numpy(dtype=np.int64).reshape(len(y)),
+        x.to_numpy(dtype=np.float64),
+        params_sr,
+    )
+
+
+def _multinomial_probit_loglikeobs(params, y, x, cov_structure, integration_method):
+    r"""Individual log-likelihood of the multinomial probit model.
+
+    .. math::
+
+    Parameters
+    ----------
+    formula : str
+              A patsy formula comprising the dependent variable and the independent variables.
+
+    y : np.array
+        1d numpy array of shape :math:'n_obs' with the observed choices
+
+    x : np.array
+        2d numpy array of shape :math:'(n_obs, n_var)' including the independent variables.
+
+    cov_structure : str
+                    Available options are 'iid' or 'free'.
+
+    integration_method : str
+                         Either 'mc_integration', 'smooth_mc_integration'
+                         or 'gauss_integration'
+
+
+    Returns:
+    --------
+        loglikeobs : np.array
+                     1d numpy array of shape :math:'(n_obs)' with
+                     the respective likelihood contribution.
+    """
+    n_var = np.shape(x)[1]
+    n_choices = len(np.unique(y))
+
+    if cov_structure == "iid":
+        cov = np.eye(n_choices - 1) + np.ones((n_choices - 1, n_choices - 1))
+
+    else:
+        covariance = params["value"]["covariance"].to_numpy()
+
+        cov = np.zeros((n_choices - 1, n_choices - 1))
+
+        a = 0
+        for i in range(n_choices - 1):
+            k = i + 1
+            cov[i, : (i + 1)] = covariance[a : (a + k)]
+            a += k
+
+        for i in range(n_choices - 1):
+            cov[i, (i + 1) :] = cov[(i + 1) :, i]
+
+    bethas = np.zeros((n_var, n_choices))
+
+    for i in range(n_choices - 1):
+        bethas[:, i] = params["value"]["choice_{}".format(i)].to_numpy()
+
+    u_prime = x.dot(bethas)
+
+    choice_prob_obs = getattr(temfpy.integration_methods, integration_method)(
+        u_prime, cov, y
+    )
+
+    choice_prob_obs[choice_prob_obs <= 1e-250] = 1e-250

```

Comment:
I think we could use [pandas.DataFrame.clip](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.clip.html) here?